rust/agents/relayer/src/msg/pending_operation.rs

```diff
@@ -0,0 +1,175 @@
+use eyre::Report;
+
+#[allow(unused_imports)] // required for enum_dispatch
+use super::pending_message::PendingMessage;
+
+/// A pending operation that will be run by the submitter and cause a
+/// transaction to be sent.
+#[async_trait]
+#[enum_dispatch]
+pub trait PendingOperation {
+    /// Prepare to run this operation. This will be called before every run and
+    /// will usually have a very short gap between it and the run call.
+    async fn prepare(&mut self) -> TxPrepareResult {
+        if self.ready_to_be_processed() {
+            TxPrepareResult::Ready
+        } else {
+            TxPrepareResult::NotReady
+        }
+    }
+
+    /// Submit this operation to the blockchain and report if it was successful
+    /// or not.
+    async fn submit(&mut self) -> TxRunResult;
+
+    async fn validate(&mut self) -> TxValidationResult {
+        // default implementation is basically a no-op
+        if self.ready_to_be_validated() {
+            TxValidationResult::Valid
+        } else if !self.submitted() {
+            TxValidationResult::Invalid
+        } else {
+            TxValidationResult::Retry
+        }
+    }
+
+    fn next_attempt_after(&self) -> Option<Instant>;
+
+    fn submitted(&self) -> bool;
+
+    fn ready_to_be_processed(&self) -> bool {
+        !self.submitted()
+            && self
+                .next_attempt_after()
+                .map_or(true, |a| Instant::now() >= a)
+    }
+
+    fn ready_to_be_validated(&self) -> bool {
+        self.submitted()
+            && self
+                .next_attempt_after()
+                .map_or(true, |a| Instant::now() >= a)
+    }
+}
+
+/// A "dynamic" pending operation implementation which knows about the
+/// different sub types and can properly implement PartialEq and
+/// PartialOrd for them.
+#[enum_dispatch(PendingOperation)]
+#[derive(Debug, PartialEq, Eq)]
+pub enum DynPendingOperation {
+    PendingMessage,
+}
+
+impl PartialOrd for DynPendingOperation {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Sort by their next allowed attempt time and if no allowed time is set,
+/// then put it in front of those with a time (they have been tried
+/// before) and break ties between ones that have not been tried with
+/// the nonce.
+impl Ord for DynPendingOperation {
+    fn cmp(&self, other: &Self) -> Ordering {
+        use DynPendingOperation::*;
+        use Ordering::*;
+        match (self.next_attempt_after(), other.next_attempt_after()) {
+            (Some(a), Some(b)) => a.cmp(&b),
+            (Some(_), None) => Greater,
+            (None, Some(_)) => Less,
+            (None, None) => match (self, other) {
+                (PendingMessage(a), PendingMessage(b)) => a.message.nonce.cmp(&b.message.nonce),
+            },
+        }
+    }
+}
+
+#[allow(dead_code)] // Inner types are for present _and_ future use so allow unused variants.
+pub enum TxPrepareResult {
```

Comment:
nit: If I'm right about validate/prepare/submit being the order, please order these enums accordingly

Reply:
It's `Prepare`, `Submit`, `Validate`. Validate makes sure it was not re-org-ed out. LMK if you have a better name for this.

Reply:
Yeah I figured this out later. Suggested "Confirm" in a different comment